subsys/jwt/jwt_ecdsa.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2018 Linaro Ltd
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/types.h>
+#include <errno.h>
+
+#include <zephyr/data/jwt.h>
+#include <zephyr/data/json.h>
+
+#if defined(CONFIG_JWT_SIGN_ECDSA_TC)
+#include <tinycrypt/ctr_prng.h>
+#include <tinycrypt/sha256.h>
+#include <tinycrypt/ecc_dsa.h>
+#include <tinycrypt/constants.h>
+
+#include <zephyr/random/random.h>
+
+static TCCtrPrng_t prng_state;
+static bool prng_init;
+
+static const char personalize[] = "zephyr:drivers/jwt/jwt.c";
+
+static int setup_prng(void)
+{
+	if (prng_init) {
+		return 0;
+	}
+	prng_init = true;
+
+	uint8_t entropy[TC_AES_KEY_SIZE + TC_AES_BLOCK_SIZE];
+
+	sys_rand_get(entropy, sizeof(entropy));
+
+	int res = tc_ctr_prng_init(&prng_state, (const uint8_t *) &entropy, sizeof(entropy),
+				   personalize, sizeof(personalize));
+
+	return res == TC_CRYPTO_SUCCESS ? 0 : -EINVAL;
+}
+
+int default_CSPRNG(uint8_t *dest, unsigned int size)
+{
+	int res = tc_ctr_prng_generate(&prng_state, NULL, 0, dest, size);
+	return res;
+}
+
+int jwt_sign_impl(struct jwt_builder *builder, const unsigned char *der_key, size_t der_key_len,
+		  unsigned char *sig, size_t sig_len)
+{
+	struct tc_sha256_state_struct ctx;
+	uint8_t hash[32];
+	int res;
+	(void) sig_len;
+
+	tc_sha256_init(&ctx);
+	tc_sha256_update(&ctx, builder->base, builder->buf - builder->base);
+	tc_sha256_final(hash, &ctx);
+
+	res = setup_prng();
+
+	if (res != 0) {
+		return res;
+	}
+	uECC_set_rng(&default_CSPRNG);
+
+	/* Note that tinycrypt only supports P-256. */
+	res = uECC_sign(der_key, hash, sizeof(hash), sig, &curve_secp256r1);
+	if (res != TC_CRYPTO_SUCCESS) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_JWT_SIGN_ECDSA_TC */
+
+#if defined(CONFIG_JWT_SIGN_ECDSA_PSA)
+#include <psa/crypto.h>
+
+int jwt_sign_impl(struct jwt_builder *builder, const unsigned char *der_key, size_t der_key_len,
+		  unsigned char *sig, size_t sig_len)
+{
+	psa_status_t status;
+	psa_key_attributes_t attr = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = 0;
+	size_t sig_len_out;
+	int ret;
+
+	psa_set_key_type(&attr, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(&attr, 256);
+	psa_set_key_algorithm(&attr, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	psa_set_key_usage_flags(&attr, PSA_KEY_USAGE_SIGN_HASH);
+
+	status = psa_import_key(&attr, der_key, der_key_len, &key_id);
+	if (status != PSA_SUCCESS) {
+		ret = -EINVAL;
+		goto exit;
+	}
+
+	status = psa_sign_hash(key_id, PSA_ALG_SHA_256, builder->base, builder->buf - builder->base,
+			       sig, sig_len, &sig_len_out);
+	if (status != PSA_SUCCESS) {
+		ret = -EINVAL;
+	}
+
+exit:
+	psa_reset_key_attributes(&attr);
```

Comment:
This [might free additional resources](https://github.com/Mbed-TLS/mbedtls/blob/8ea8b30941a6d364f5bf4b3850f5312112e760d9/tf-psa-crypto/include/psa/crypto.h#L391) allocated with the attributes, but after checking the current implementation in Mbed TLS, there's no extra memory allocated with `psa_key_attributes_t` so `psa_reset_key_attributes` is not really useful indeed

Reply:
Yeah. Well, because of how it's documented, then any way is fine by me. I imagine the unnecessary `psa_reset_key_attributes()` would be optimized out by the compiler if it does nothing interesting.